pkg/prove/querier.go

```diff
@@ -54,3 +58,92 @@ func QueryTxInclusionProof(_ sdk.Context, path []string, req abci.RequestQuery)
 
 	return rawTxProof, nil
 }
+
+const ShareInclusionQueryPath = "shareInclusionProof"
+
+// QueryShareInclusionProof defines the logic performed when querying for the
+// inclusion proofs of a set of shares to the data root. The share range should
+// be appended to the path. Example path for proving the set of shares [3, 5]:
+// custom/shareInclusionProof/3/5
+func QueryShareInclusionProof(_ sdk.Context, path []string, req abci.RequestQuery) ([]byte, error) {
+	// parse the share range from the path
+	if len(path) != 2 {
+		return nil, fmt.Errorf("expected query path length: 2 actual: %d ", len(path))
+	}
+	beginShare, err := strconv.ParseInt(path[0], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	endShare, err := strconv.ParseInt(path[1], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	// unmarshal the block data that is passed from the ABCI client
+	pbb := new(tmproto.Block)
+	err = pbb.Unmarshal(req.Data)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block: %w", err)
+	}
+	data, err := types.DataFromProto(&pbb.Data)
+	if err != nil {
+		panic(fmt.Errorf("error from proto block: %w", err))
+	}
+
+	rawShares, err := shares.Split(data, true)
+	if err != nil {
+		return nil, err
+	}
+
+	nID, err := ParseNamespaceID(rawShares, beginShare, endShare)
+	if err != nil {
+		return nil, err
+	}
+
+	// create and marshal the shares inclusion proof, which we return in the form of []byte
+	txProof, err := SharesInclusion(
+		rawShares,
+		data.SquareSize,
+		nID,
+		uint64(beginShare),
+		uint64(endShare),
+	)
+	if err != nil {
+		return nil, err
+	}
+	pTxProof := txProof.ToProto()
+	rawTxProof, err := pTxProof.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	return rawTxProof, nil
+}
+
+// ParseNamespaceID validates the shares range and returns their namespace ID.
+func ParseNamespaceID(rawShares []shares.Share, startShare int64, endingShare int64) (namespace.ID, error) {
```

Comment:
[nit] for consistency, can both params please use the same state/tense?

```suggestion
func ParseNamespaceID(rawShares []shares.Share, startShare int64, endShare int64) (namespace.ID, error) {
```

alternatively: `startingShare` and `endingShare`

Accepting this proposal implies correcting the error messages below

Reply:
[508e24f](https://github.com/celestiaorg/celestia-app/pull/1233/commits/508e24f194b4b67f98775f388fa41a423d7364ec)